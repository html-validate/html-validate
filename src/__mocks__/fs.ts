import { fs as memfs } from "memfs";

const fs = {
	__esModule: true,
	existsSync: memfs.existsSync.bind(memfs),
	lstatSync: memfs.lstatSync.bind(memfs),
	mkdirSync: memfs.mkdirSync.bind(memfs),
	readFileSync: memfs.readFileSync.bind(memfs),
	readdir: memfs.readdir.bind(memfs),
	readdirSync: memfs.readdirSync.bind(memfs),
	readlinkSync: memfs.readlinkSync.bind(memfs),
	realpath: memfs.realpath.bind(memfs),
	realpathSync: { native: memfs.realpathSync.bind(memfs) },
	statSync: memfs.statSync.bind(memfs),
	writeFile: memfs.writeFile.bind(memfs),
	writeFileSync: memfs.writeFileSync.bind(memfs),
};

const proxy = new Proxy(fs, {
	get(target, prop, receiver): any {
		if (prop === "default") {
			return proxy;
		}
		if (Reflect.has(target, prop)) {
			return Reflect.get(target, prop, receiver);
		} else {
			throw new Error(`fs mock does not implement "${prop.toString()}"`);
		}
	},
});

module.exports = proxy;
