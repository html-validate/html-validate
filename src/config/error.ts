import { UserError } from "../error";

/**
 * @public
 */
export class ConfigError extends UserError {
	public constructor(message: string, nested?: Error) {
		super(message, nested);
		Error.captureStackTrace(this, ConfigError);
		this.name = ConfigError.name;
	}
}
