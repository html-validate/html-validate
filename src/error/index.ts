export { UserError } from "./user-error";
export { NestedError } from "./nested-error";
export { SchemaValidationError } from "./schema-validation-error";
