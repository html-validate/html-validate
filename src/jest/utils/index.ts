export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
