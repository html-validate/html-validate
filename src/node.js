'use strict';

class Node {
	constructor(tagName, parent){
		this.children = [];
		this.tagName = tagName;
		this.parent = parent;
		this.attr = {};
	}

	static rootNode(){
		return new Node();
	}

	static fromToken(token, parent, config){
		let node = new Node(token.data[2], parent);
		node.selfClosed = !!token.data[3];
		node.voidElement = Node.isVoidElement(config, node.tagName);
		return node;
	}

	static isVoidElement(config, tagName){
		return config.html.voidElements.indexOf(tagName.toLowerCase()) !== -1;
	}

	setAttribute(key, value){
		this.attr[key] = value;
	}

	getAttribute(key){
		return this.attr[key];
	}

	append(node){
		this.children.push(node);
	}
}

module.exports = Node;
