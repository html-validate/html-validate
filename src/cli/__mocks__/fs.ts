import path from "path";
import { Stats } from "fs";

interface Entry {
	type: "file" | "directory";
}

interface FileEntry extends Entry {
	type: "file";
	content: string;
}

interface DirectoryEntry extends Entry {
	type: "directory";
}

class FSError extends Error {
	public code: string;
	public path: string;

	public constructor(code: string, message: string, path: string) {
		super(message);
		this.code = code;
		this.path = path;
	}

	public toString(): string {
		return `${this.code}: ${this.message} '${this.path}'`;
	}
}

class ENOENT extends FSError {
	public constructor(filePath: string) {
		super("ENOENT", "no such file or directory", filePath);
	}
}

let mockData: Map<string, Entry> = new Map();

function isDirectory(entry: Entry): entry is DirectoryEntry {
	return entry.type === "directory";
}

function isFile(entry: Entry): entry is FileEntry {
	return entry.type === "file";
}

function normalize(filePath: string): string {
	return path.isAbsolute(filePath) ? filePath : path.join(process.cwd(), filePath);
}

function mockFile(filePath: string, content: string): void {
	const normalized = normalize(filePath);
	const fileEntry: FileEntry = {
		type: "file",
		content,
	};
	mockData.set(normalized, fileEntry);
	let current = path.dirname(normalized);
	while (current !== "/") {
		if (mockData.has(current)) {
			break;
		}
		const dirEntry: DirectoryEntry = {
			type: "directory",
		};
		mockData.set(current, dirEntry);
		current = path.dirname(current);
	}
}

function mockReset(): void {
	mockData = new Map();
	const rootEntry: DirectoryEntry = {
		type: "directory",
	};
	mockData.set("/", rootEntry);
}

function existsSync(filePath: string): boolean {
	return mockData.has(normalize(filePath));
}

function readdirSync(
	filePath: string,
	options?: string | { encoding?: string; withFileTypes?: boolean }
): string[] | Buffer[] {
	if (!filePath) {
		throw new ENOENT(filePath);
	}
	const normalized = normalize(filePath);
	const entry = mockData.get(normalized);
	if (!entry) {
		throw new ENOENT(filePath);
	}
	if (!isDirectory(entry)) {
		throw new FSError("ENOTDIR", "not a directory", filePath);
	}
	const encoding = typeof options === "string" ? options : options?.encoding ?? "utf-8";
	const entries = Array.from(mockData.keys())
		.filter((it) => path.dirname(it) === normalized)
		.map((it) => path.basename(it))
		.sort();
	if (typeof options !== "string" && options?.withFileTypes) {
		throw new Error("NotImplemented: readdirSync option withFileTypes is not supported by mock");
	}
	if (encoding !== "utf-8") {
		throw new Error("NotImplemented: readdirSync encoding only supports utf-8");
	}
	return entries;
}

function readFileSync(
	filePath: string,
	options?: string | { encoding?: string; flag?: string }
): Buffer | string {
	if (!filePath) {
		throw new ENOENT(filePath);
	}
	const normalized = normalize(filePath);
	const entry = mockData.get(normalized);
	if (!entry) {
		throw new ENOENT(filePath);
	}
	if (!isFile(entry)) {
		throw new FSError("EISDIR", "illegal operation on a directory", filePath);
	}
	if (typeof options === "string" || options?.encoding) {
		return entry.content;
	} else {
		return Buffer.from(entry.content, "utf-8");
	}
}

function statSync(filePath: string): Stats {
	if (!filePath) {
		throw new ENOENT(filePath);
	}
	const normalized = normalize(filePath);
	const date = new Date();
	const time = date.getTime();
	const entry = mockData.get(normalized);
	if (!entry) {
		throw new ENOENT(filePath);
	}
	return {
		isFile: () => isFile(entry),
		isDirectory: () => isDirectory(entry),
		isBlockDevice: () => false,
		isCharacterDevice: () => false,
		isSymbolicLink: () => false,
		isFIFO: () => false,
		isSocket: () => false,
		dev: 1,
		ino: 1,
		mode: 0b110110110,
		nlink: 1,
		uid: 1,
		gid: 1,
		rdev: 1,
		size: 1,
		blksize: 1,
		blocks: 1,
		atimeMs: time,
		mtimeMs: time,
		ctimeMs: time,
		birthtimeMs: time,
		atime: date,
		mtime: date,
		ctime: date,
		birthtime: date,
	};
}

const original = jest.requireActual("fs");
const fs = {
	__esModule: undefined,

	mockFile,
	mockReset,

	existsSync,
	realpath: original.realpath,
	realpathSync: original.realpathSync,
	readdirSync,
	readFileSync,
	statSync,
	lstatSync: statSync,
};

module.exports = new Proxy(fs, {
	get(target, prop, receiver) {
		if (Reflect.has(target, prop)) {
			return Reflect.get(target, prop, receiver);
		} else {
			throw new Error(`fs mock does not implement "${prop.toString()}"`);
		}
	},
});
