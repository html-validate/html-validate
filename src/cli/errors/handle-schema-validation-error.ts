import fs from "node:fs";
import path from "node:path";
import betterAjvErrors from "@sidvind/better-ajv-errors";
import kleur from "kleur";
import { type SchemaValidationError } from "../../error";

function prettyError(err: SchemaValidationError): string {
	let json: string | undefined;
	if (err.filename && fs.existsSync(err.filename)) {
		json = fs.readFileSync(err.filename, "utf-8");
	}
	return betterAjvErrors(err.schema, err.obj, err.errors, {
		format: "cli",
		indent: 2,
		json,
	});
}

/**
 * @internal
 */
export function handleSchemaValidationError(console: Console, err: SchemaValidationError): void {
	if (err.filename) {
		const filename = path.relative(process.cwd(), err.filename);
		console.error(kleur.red(`A configuration error was found in "${filename}":`));
	} else {
		console.error(kleur.red(`A configuration error was found:`));
	}
	console.group();
	{
		console.error(prettyError(err));
	}
	console.groupEnd();
}
