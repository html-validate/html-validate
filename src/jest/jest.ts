import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
