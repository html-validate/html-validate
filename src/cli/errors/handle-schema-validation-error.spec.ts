import fs from "node:fs";
import path from "node:path";
import { Console } from "node:console";
import { WritableStreamBuffer } from "stream-buffers";
import Ajv, { type SchemaObject } from "ajv";
import kleur from "kleur";
import { SchemaValidationError } from "../../error";
import { stripAnsi } from "../../jest/utils";
import { ajvFunctionKeyword, ajvRegexpKeyword } from "../../schema/keywords";
import { handleSchemaValidationError } from "./handle-schema-validation-error";

kleur.enabled = true;

let stdout: WritableStreamBuffer;
let stderr: WritableStreamBuffer;
let console: Console;

beforeEach(() => {
	stdout = new WritableStreamBuffer();
	stderr = new WritableStreamBuffer();
	console = new Console(stdout, stderr);
});

function loadFixture(filename: string): { data: unknown; filePath: string } {
	const filePath = path.join(__dirname, "__fixtures__", filename);
	return { data: JSON.parse(fs.readFileSync(filePath, "utf-8")), filePath };
}

function errorFromFixture(
	filename: string | null,
	data: unknown,
	schema: SchemaObject,
): SchemaValidationError {
	const ajv = new Ajv({ strict: true, strictTuples: true, strictTypes: true });
	ajv.addKeyword(ajvFunctionKeyword);
	ajv.addKeyword(ajvRegexpKeyword);
	const validator = ajv.compile(schema);
	validator(data);
	const errors = validator.errors ?? [];
	return new SchemaValidationError(filename, "Mock error", data, schema, errors);
}

it("should present error in pretty format", () => {
	expect.assertions(2);
	const { data, filePath } = loadFixture("invalid.json");
	const error = errorFromFixture(filePath, data, {
		type: "object",
		properties: {
			foo: {
				type: "string",
			},
		},
	});
	handleSchemaValidationError(console, error);
	expect(stripAnsi(stdout.getContentsAsString("utf-8"))).toMatchSnapshot("stdout");
	expect(stripAnsi(stderr.getContentsAsString("utf-8"))).toMatchSnapshot("stderr");
});

it("should handle when filename is missing", () => {
	expect.assertions(2);
	const { data } = loadFixture("invalid.json");
	const error = errorFromFixture(null, data, {
		type: "object",
		properties: {
			foo: {
				type: "string",
			},
		},
	});
	handleSchemaValidationError(console, error);
	expect(stripAnsi(stdout.getContentsAsString("utf-8"))).toMatchSnapshot("stdout");
	expect(stripAnsi(stderr.getContentsAsString("utf-8"))).toMatchSnapshot("stderr");
});

it("should handle when filename does not exist", () => {
	expect.assertions(2);
	const { data } = loadFixture("invalid.json");
	const error = errorFromFixture("missing-file.json", data, {
		type: "object",
		properties: {
			foo: {
				type: "string",
			},
		},
	});
	handleSchemaValidationError(console, error);
	expect(stripAnsi(stdout.getContentsAsString("utf-8"))).toMatchSnapshot("stdout");
	expect(stripAnsi(stderr.getContentsAsString("utf-8"))).toMatchSnapshot("stderr");
});

it("should handle function keyword", () => {
	expect.assertions(2);
	const { data, filePath } = loadFixture("invalid.json");
	const error = errorFromFixture(filePath, data, {
		type: "object",
		properties: {
			foo: {
				function: true,
			},
		},
	});
	handleSchemaValidationError(console, error);
	expect(stripAnsi(stdout.getContentsAsString("utf-8"))).toMatchSnapshot("stdout");
	expect(stripAnsi(stderr.getContentsAsString("utf-8"))).toMatchSnapshot("stderr");
});

it("should handle regexp keyword", () => {
	expect.assertions(2);
	const { data, filePath } = loadFixture("invalid.json");
	const error = errorFromFixture(filePath, data, {
		type: "object",
		properties: {
			foo: {
				regexp: true,
			},
		},
	});
	handleSchemaValidationError(console, error);
	expect(stripAnsi(stdout.getContentsAsString("utf-8"))).toMatchSnapshot("stdout");
	expect(stripAnsi(stderr.getContentsAsString("utf-8"))).toMatchSnapshot("stderr");
});
