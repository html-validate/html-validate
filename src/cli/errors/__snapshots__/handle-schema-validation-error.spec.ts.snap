// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should handle function keyword: stderr 1`] = `
"A configuration error was found in "src/cli/errors/__fixtures__/invalid.json":
  TYPE should be a function
  
    1 | {
  > 2 |   "foo": [1, 2, 3],
      |          ^^^^^^^^^ type should be a function
    3 |   "bar": "baz"
    4 | }
    5 |
"
`;

exports[`should handle function keyword: stdout 1`] = `""`;

exports[`should handle regexp keyword: stderr 1`] = `
"A configuration error was found in "src/cli/errors/__fixtures__/invalid.json":
  TYPE should be a regular expression
  
    1 | {
  > 2 |   "foo": [1, 2, 3],
      |          ^^^^^^^^^ type should be a regular expression
    3 |   "bar": "baz"
    4 | }
    5 |
"
`;

exports[`should handle regexp keyword: stdout 1`] = `""`;

exports[`should handle when filename does not exist: stderr 1`] = `
"A configuration error was found in "missing-file.json":
  TYPE must be string
  
    1 | {
  > 2 |   "foo": [
      |          ^
  > 3 |     1,
      | ^^^^^^
  > 4 |     2,
      | ^^^^^^
  > 5 |     3
      | ^^^^^^
  > 6 |   ],
      | ^^^^ type must be string
    7 |   "bar": "baz"
    8 | }
"
`;

exports[`should handle when filename does not exist: stdout 1`] = `""`;

exports[`should handle when filename is missing: stderr 1`] = `
"A configuration error was found:
  TYPE must be string
  
    1 | {
  > 2 |   "foo": [
      |          ^
  > 3 |     1,
      | ^^^^^^
  > 4 |     2,
      | ^^^^^^
  > 5 |     3
      | ^^^^^^
  > 6 |   ],
      | ^^^^ type must be string
    7 |   "bar": "baz"
    8 | }
"
`;

exports[`should handle when filename is missing: stdout 1`] = `""`;

exports[`should present error in pretty format: stderr 1`] = `
"A configuration error was found in "src/cli/errors/__fixtures__/invalid.json":
  TYPE must be string
  
    1 | {
  > 2 |   "foo": [1, 2, 3],
      |          ^^^^^^^^^ type must be string
    3 |   "bar": "baz"
    4 | }
    5 |
"
`;

exports[`should present error in pretty format: stdout 1`] = `""`;
